pkg/cmd/pulumi/new_ai.go

```diff
@@ -0,0 +1,263 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+
+	survey "github.com/AlecAivazis/survey/v2"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/env"
+)
+
+const (
+	noSelection     = "no"
+	yesSelection    = "yes"
+	refineSelection = "refine"
+)
+
+func deriveAIOrTemplate(args newArgs) string {
+	if args.aiPrompt != "" || args.aiLanguage != "" {
+		return "ai"
+	}
+	return "template"
+}
+
+func shouldPromptForAIOrTemplate(args newArgs, userBackend backend.Backend) bool {
+	_, isHTTPBackend := userBackend.(httpstate.Backend)
+	return args.aiPrompt == "" &&
+		args.aiLanguage == "" &&
+		!args.templateMode &&
+		isHTTPBackend
+}
+
+// Iteratively prompt the user for input, sending their input as a prompt tp Pulumi AI
+// Stream the response back to the console, and repeat until the user is done.
+func runAINew(
+	ctx context.Context,
+	args newArgs,
+	opts display.Options,
+	backend httpstate.Backend,
+) (conversationURL string, err error) {
+	languageOptions := []string{
+		"TypeScript",
+		"JavaScript",
+		"Python",
+		"Go",
+		"C#",
+		"Java",
+		"YAML",
+	}
+	var rawLanguageSelect string
+	if args.aiLanguage == "" {
+		if err = survey.AskOne(&survey.Select{
+			Message: "Please select a language for your project:",
+			Options: languageOptions,
+		}, &rawLanguageSelect, surveyIcons(opts.Color)); err != nil {
+			return "", err
+		}
+		err = args.aiLanguage.Set(rawLanguageSelect)
+		if err != nil {
+			return "", err
+		}
+	}
+	var continuePrompt string
+	var connectionID string
+	var conversationID string
+	for continuePrompt, conversationURL, conversationID, connectionID, err = runAINewPromptStep(
+		ctx,
+		backend,
+		opts,
+		args,
+		"",
+		"",
+		"",
+		"",
+	); continuePrompt == refineSelection; continuePrompt,
+		conversationURL,
+		conversationID,
+		connectionID,
+		err = runAINewPromptStep(
+		ctx,
+		backend,
+		opts,
+		args,
+		continuePrompt,
+		connectionID,
+		conversationURL,
+		conversationID,
+	) {
+		if err != nil {
+			return "", err
+		}
+	}
+	return conversationURL, err
+}
+
+func sendPromptToPulumiAI(
+	ctx context.Context,
+	backend httpstate.Backend,
+	promptMessage string,
+	conversationID string,
+	connectionID string,
+	language httpstate.PulumiAILanguage,
+) (string, string, string, error) {
+	pulumiAIURL := env.AIServiceEndpoint.Value()
+	if pulumiAIURL == "" {
+		pulumiAIURL = "https://www.pulumi.com/ai"
+	}
+	parsedURL, err := url.Parse(pulumiAIURL)
+	if err != nil {
+		return "", "", "", err
+	}
```

Comment:
Should the URL somehow be encapsulated within the backend? Is it always going to be `https://www.pulumi.com/ai`, even with self-hosted backends?

Reply:
Yes - the zip URLs are located there presently. We don't expect self-hosted to be supported for a while yet after this release as we need to address other complexity around hosting Pulumi AI.